rfcs/2021-07-27-7117-vrl-string-interpolation.md

```diff
@@ -0,0 +1,182 @@
+# RFC 7117 - 2021-07-27 - VRL string interpolation
+
+VRL needs a better way to format strings. Currently the only way to do this is
+to concatenate strings, which can get unwieldly.
+
+## Scope
+
+This RFC discusses creating a new string type within VRL that can use template
+literals, also known as string interpolation, to format strings.
+
+## Pain
+
+Currently the way to create strings is through either string concatenation or
+the `join` function.
+
+Syntactically this is unwieldly. It requires extra key presses and the code
+created doesn't necessarily give an instant idea of what the resulting string
+will look like. Thus the true intent behind the code is obfuscated, which can
+result in bugs.
+
+## User Experience
+
+We will be loosely basing our format strings on Pythons [f-strings](https://peps.python.org/pep-0498/).
+
+To format a string there will be a string type denoted with the prefix `f'`.
+
+Within that string, it is possible to embed VRL expressions by surrounding them
+with `{..}`. VRL will evaluate the expressions and will call `as_string` on
+that expression to return the text representation.
+
+```coffee
+f'The message is { .message } created at { .timestamp }'
+```
+
+If you wish to actually insert a `{` into a string, a double '{{' will be needed.
+
+```coffee
+f'Here is a curly brace -> {{'
+```
+
+Since this is a new string type there are no backward compatibility issues.
+
+### Format strings
+
+The format can be specified by adding format strings after a `:` in the string.
+
+For example to format date the following would be valid:
+
+```coffee
+ts = t'2020-10-21T16:00:00Z'
+f'The time is {ts : %v %R}'
+```
+
+### Errors
+
+The question arises about what we should do if an error occurs.
```

Comment:
I think this is a pretty central question to this PR where we want resolution before going ahead?

Reply:
Yes. 

Personally I think it best for user experience to make it infallible, but others may disagree.

Reply:
I agree if we can use VRL's type system to take the format string into account 😄